teuthology/task/kernel.py

```diff
@@ -1230,105 +1333,12 @@ def task(ctx, config):
-                )
-                msg = msg.format(system_type=system_type)
-                log.error(msg)
-                ctx.summary["failure_reason"] = msg
-                ctx.summary["status"] = "dead"
-                raise ConfigError(msg)
-
-            # FIXME: this install should probably happen somewhere else
-            # but I'm not sure where, so we'll leave it here for now.
-            install_package('koji', role_remote)
-
-            if build_id:
-                # get information about this build from koji
-                build_info = get_koji_build_info(build_id, role_remote, ctx)
-                version = "{ver}-{rel}.x86_64".format(
-                    ver=build_info["version"],
-                    rel=build_info["release"]
-                )
-            elif task_id:
-                # get information about results of this task from koji
-                task_result = get_koji_task_result(task_id, role_remote, ctx)
-                # this is not really 'build_info', it's a dict of information
-                # about the kernel rpm from the task results, but for the sake
-                # of reusing the code below I'll still call it that.
-                build_info = get_koji_task_rpm_info(
-                    'kernel',
-                    task_result['rpms']
-                )
-                # add task_id so we can know later that we're installing
-                # from a task and not a build.
-                build_info["task_id"] = task_id
-                version = build_info["version"]
-
-            if need_to_install(ctx, role, version):
-                need_install[role] = build_info
-                need_version[role] = version
-        else:
-            builder = get_builder_project()(
-                "kernel",
-                role_config,
-                ctx=ctx,
-                remote=role_remote,
-            )
-            sha1 = builder.sha1
-            log.debug('sha1 for {role} is {sha1}'.format(role=role, sha1=sha1))
-            ctx.summary['{role}-kernel-sha1'.format(role=role)] = sha1
-
-            if need_to_install(ctx, role, sha1):
-                if teuth_config.use_shaman:
-                    version = builder.scm_version
-                else:
-                    version = builder.version
-                if not version:
-                    raise VersionNotFoundError(builder.base_url)
-                need_install[role] = sha1
-                need_version[role] = version
-
-        # enable or disable kdb if specified, otherwise do not touch
-        if role_config.get('kdb') is not None:
-            kdb[role] = role_config.get('kdb')
-
-    if need_install:
-        install_firmware(ctx, need_install)
-        download_kernel(ctx, need_install)
-        install_and_reboot(ctx, need_install, config)
-        wait_for_reboot(ctx, need_version, timeout, config)
+        t = threading.Thread(target=process_role, args=(ctx, config, timeout,
```

Comment:
you can use the parallel() construction to simplify this a little, and use gevent rather than native threads, e..g. https://github.com/ceph/teuthology/blob/master/teuthology/reimage.py#L54 - looks good though!